### ISSUE TITLE
[HELP] Convert Rust struct to Java object

### ISSUE THREAD

**Issue by Totobird-Creations** [astonbitecode/j4rs#38]
I have the following piece of code, but it fails to compile with the message: `the trait bound 'InvocationArg: From<test_object::TestObject>' is not satisfied`.
```rust
#[derive(Serialize, Deserialize)]
pub struct TestObject {
    content : String
}
fn create(text : Instance /*String*/) -> Result<Instance /*TestObject*/, String> {
    // Argument java-to-rust conversion omitted.
    let instance = TestObject {
        content : text
    }
    let ia = InvocationArg::try_from(instance).map_err(|error| format!("{}", error))?;
    return Instance::try_from(ia).map_err(|error| format!("{}", error))?;
}
```
I tried implementing it with the following, but the java side fails when I call `Java2RustUtils.getObjectCasted(...)`, stating that the object was created in Rust.
```rust
impl TryFrom<TestObject> for InvocationArg {
    type Error = errors::J4RsError;
    fn try_from(arg : TestObject) -> errors::Result<InvocationArg> {
        return Ok(InvocationArg::new(&arg, "io.example.TestObject"));
    }
}
```
How can I convert the Rust struct to a Java object, or is there something else that I can do that is consistent with how I could convert the default types (`String`, `i32`, `f64`, etc)?

**Comment by astonbitecode** [comment]
Your `TestObject` in the Rust world should have a respective `TestObject` declaration in the Java world:

```rust
#[derive(Serialize, Deserialize)]
pub struct TestObject {
    content : String
}
```

```java
package my.tests;

public class TestObject {
    public String content;
	
	public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }
```

You can create an `InvocationArg` in Rust like following

```rust
let test_object = TestObject {
	content : text
};
let ia = InvocationArg::new(&test_object, "my.tests.TestObject");
return Instance::try_from(ia).map_err(|error| format!("{}", error))?;
```

The Java world _should_ create a respective `TestObject` instance, which you could get in the Java world like:

```java
Java2RustUtils.getObjectCasted(instance);
```

However, even if using custom structs/classes was always supported for Rust -> Java direction (passing as arguments to Java methods) it seems that this specific is a feature that was missed during the Java -> Rust direction implementation, which started in `j4rs` v0.12.0. Currently, the `getObjectCasted` does not work as expected.

A temporary solution to get your case working with 0.13.0 would be to create the object in the Java world explicitly, instead of doing implicit transformation from struct to class:

```rust
let test_object_instance = jvm.create_instance("my.tests.TestObject", &[]).unwrap();
let invocation_args = vec![InvocationArg::try_from("Some content").unwrap()];
let _ = jvm.invoke(&test_object_instance, "setContent", &invocation_args);

return Ok(test_object_instance)
```

I will implement the needed `Java2RustUtils` behavior shortly.

Thanks for reporting!

**Comment by astonbitecode** [comment]
Ok the implementation is now in the master. I added the case as example [here](https://github.com/astonbitecode/j4rs-java-call-rust/blob/33a3d54951ed90390aa10ab9a8d49fc8ee8b8920/java/src/main/java/io/github/astonbitecode/j4rs/example/RustFunctionCalls.java#L35).

Until it is released, you may try it, using the `j4rs` [0.13.1-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/io/github/astonbitecode/j4rs/0.13.1-SNAPSHOT/).

**Comment by astonbitecode** [comment]
Released (v0.14.0)
